app/forms/silo-create.tsx

```diff
@@ -65,7 +66,22 @@ export function CreateSiloSideModalForm() {
   })
 
   const form = useForm({ defaultValues })
-
+  const identityMode = form.watch('identityMode')
+  const adminGroupName = form.watch('adminGroupName')
+  // Clear the adminGroupName if the user selects the "local only" identity mode
+  useEffect(() => {
+    if (identityMode === 'local_only') {
+      form.setValue('adminGroupName', '')
+    }
+  }, [identityMode, form])
+  // Clear the role assignment checkboxes if the adminGroupName is deleted
+  const fleetRolesDisabled = !adminGroupName
+  useEffect(() => {
+    if (fleetRolesDisabled) {
+      form.setValue('siloAdminGetsFleetAdmin', false)
+      form.setValue('siloViewerGetsFleetViewer', false)
+    }
+  }, [fleetRolesDisabled, form])
```

Comment:
I tried doing this stuff in `onChange` on identity mode and admin group name instead, which is vaguely to be preferred over `useEffect` when you can make it work, because cuts out a render or two that have to happen while each of the state values update in turn and the useEffects fire. But it didn't really work out.

The result was a bit sad in that a) I lost the ability to clear the checkboxes when the group name clears (because `setValue` doesn't trigger `onChange`), and b) it requires a change to how we handle `onChange` in the text field and radio field that I am very unsure about. I had to make so when you pass in `onChange`, it runs in addition to the `field` one instead of overriding it. There's a good chance we are relying on that overriding behavior, plus it would be weird to have some fields work this way and others not. I'm sure the `watch` and `useEffect` version won't cause perf problems anyway, plus it's clearer because the related logic is all together.

<details>
<summary>failed experiment diff</summary>

```diff
diff --git a/app/components/form/fields/RadioField.tsx b/app/components/form/fields/RadioField.tsx
index 1fdc9c50..f2f7ece3 100644
--- a/app/components/form/fields/RadioField.tsx
+++ b/app/components/form/fields/RadioField.tsx
@@ -71,6 +71,7 @@ export function RadioField<
   control,
   items,
   parseValue,
+  onChange,
   ...props
 }: RadioFieldProps<TFieldValues, TName>) {
   const id = useId()
@@ -92,9 +93,11 @@ export function RadioField<
           [`${id}-help-text`]: !!tooltipText,
         })}
         aria-describedby={tooltipText ? `${id}-label-tip` : undefined}
-        onChange={(e) =>
-          parseValue ? field.onChange(parseValue(e.target.value)) : field.onChange(e)
-        }
+        onChange={(e) => {
+          // if onChange is passed in, call it in addition to field one, not instead of it
+          onChange?.(e)
+          return parseValue ? field.onChange(parseValue(e.target.value)) : field.onChange(e)
+        }}
         name={field.name}
         {...props}
         // TODO: once we get rid of the other use of RadioGroup, change RadioGroup
diff --git a/app/components/form/fields/TextField.tsx b/app/components/form/fields/TextField.tsx
index 6662fa07..5dff4aac 100644
--- a/app/components/form/fields/TextField.tsx
+++ b/app/components/form/fields/TextField.tsx
@@ -114,12 +114,13 @@ export const TextFieldInner = <
   required,
   id: idProp,
   transform,
+  onChange,
   ...props
 }: TextFieldProps<TFieldValues, TName> & UITextAreaProps) => {
   const generatedId = useId()
   const id = idProp || generatedId
   const {
-    field: { onChange, ...fieldRest },
+    field: { onChange: fieldOnChange, ...fieldRest },
     fieldState: { error },
   } = useController({ name, control, rules: { required, validate } })
   return (
@@ -131,7 +132,10 @@ export const TextFieldInner = <
         error={!!error}
         aria-labelledby={cn(`${id}-label`, !!tooltipText && `${id}-help-text`)}
         aria-describedby={tooltipText ? `${id}-label-tip` : undefined}
-        onChange={(e) => onChange(transform ? transform(e.target.value) : e.target.value)}
+        onChange={(e) => {
+          onChange?.(e)
+          return fieldOnChange(transform ? transform(e.target.value) : e.target.value)
+        }}
         {...fieldRest}
         {...props}
       />
diff --git a/app/forms/silo-create.tsx b/app/forms/silo-create.tsx
index a677b2a7..5b73cae0 100644
--- a/app/forms/silo-create.tsx
+++ b/app/forms/silo-create.tsx
@@ -68,20 +68,8 @@ export function CreateSiloSideModalForm() {
   const form = useForm({ defaultValues })
   const identityMode = form.watch('identityMode')
   const adminGroupName = form.watch('adminGroupName')
-  // Clear the adminGroupName if the user selects the "local only" identity mode
-  useEffect(() => {
-    if (identityMode === 'local_only') {
-      form.setValue('adminGroupName', '')
-    }
-  }, [identityMode, form])
-  // Clear the role assignment checkboxes if the adminGroupName is deleted
   const fleetRolesDisabled = !adminGroupName
-  useEffect(() => {
-    if (fleetRolesDisabled) {
-      form.setValue('siloAdminGetsFleetAdmin', false)
-      form.setValue('siloViewerGetsFleetViewer', false)
-    }
-  }, [fleetRolesDisabled, form])
+
   return (
     <SideModalForm
       form={form}
@@ -160,6 +148,14 @@ export function CreateSiloSideModalForm() {
           { value: 'saml_jit', label: 'SAML' },
           { value: 'local_only', label: 'Local only' },
         ]}
+        onChange={(e) => {
+          // Clear SAML-specific fields when user selects "local only"
+          if (e.target.value === 'local_only') {
+            form.setValue('adminGroupName', '')
+            form.setValue('siloAdminGetsFleetAdmin', false)
+            form.setValue('siloViewerGetsFleetViewer', false)
+          }
+        }}
       />
       {identityMode === 'saml_jit' && (
         <>
```

</details>